File: apps/tool-gateway/src/routes/knowledge.ts
/**
 * Knowledge Tool Routes - RAG Evidence Retrieval
 * 
 * Implements:
 * - knowledge.search: Hybrid search for evidence chunks
 * - knowledge.get_chunk: Get full chunk with citation
 * - knowledge.index_product: Index product content
 * - knowledge.sync_xoobay: Sync XOOBAY products to RAG
 */

import type { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify';
import { createSuccessResponse, createErrorResponse, createLogger, query, queryOne } from '@shopping-agent/common';
import { getXOOBAYClient } from '../services/xoobay.js';

const logger = createLogger('knowledge');

// ============================================================
// Types
// ============================================================

interface EvidenceChunk {
  id: string;
  text: string;
  source_type: string;
  offer_id: string | null;
  sku_id: string | null;
  category_id: string | null;
  language: string;
  offsets: { start: number; end: number } | null;
  created_at: Date;
}

// ============================================================
// Helper Functions
// ============================================================

/**
 * Generate simple embedding for text (TF-IDF-like approach)
 * In production, use OpenAI embeddings or similar
 */
function generateSimpleEmbedding(text: string): number[] {
  const tokens = text.toLowerCase()
    .replace(/[^\w\s]/g, ' ')
    .split(/\s+/)
    .filter(t => t.length > 2);
  
  const embedding = new Array(1536).fill(0);
  
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i];
    let hash = 0;
    for (let j = 0; j < token.length; j++) {
      hash = ((hash << 5) - hash) + token.charCodeAt(j);
      hash = hash & hash;
    }
    
    for (let k = 0; k < 4; k++) {
      const idx = Math.abs((hash + k * 17) % 1536);
      embedding[idx] += 1.0 / (1 + i * 0.1);
    }
  }
  
  const magnitude = Math.sqrt(embedding.reduce((sum, val) => sum + val * val, 0));
  if (magnitude > 0) {
    for (let i = 0; i < embedding.length; i++) {
      embedding[i] /= magnitude;
    }
  }
  
  return embedding;
}

/**
 * Chunk text into smaller pieces
 */
function chunkText(text: string, maxSize = 500): Array<{ text: string; start: number; end: number }> {
  const chunks: Array<{ text: string; start: number; end: number }> = [];
  
  if (!text || text.length === 0) return chunks;
  
  const sentences = text.split(/(?<=[.!?])\s+/);
  let currentChunk = '';
  let currentStart = 0;
  let position = 0;
  
  for (const sentence of sentences) {
    if (currentChunk.length + sentence.length > maxSize && currentChunk.length > 0) {
      chunks.push({
        text: currentChunk.trim(),
        start: currentStart,
        end: position,
      });
      currentChunk = sentence;
      currentStart = position;
    } else {
      if (currentChunk.length === 0) currentStart = position;
      currentChunk += (currentChunk.length > 0 ? ' ' : '') + sentence;
    }
    position += sentence.length + 1;
  }
  
  if (currentChunk.trim().length > 0) {
    chunks.push({
      text: currentChunk.trim(),
      start: currentStart,
      end: position,
    });
  }
  
  return chunks;
}

/**
 * Generate citation
 */
function generateCitation(chunk: EvidenceChunk): string {
  const parts: string[] = [];
  if (chunk.offer_id) parts.push(`offer:${chunk.offer_id}`);
  if (chunk.source_type) parts.push(`type:${chunk.source_type}`);
  parts.push(`chunk:${chunk.id}`);
  return `[${parts.join(', ')}]`;
}

// ============================================================
// Routes
// ============================================================

export async function knowledgeRoutes(app: FastifyInstance): Promise<void> {
  /**
   * knowledge.search
   * 
   * Hybrid search for evidence chunks (keyword + semantic)
   */
  app.post('/search', async (request: FastifyRequest, reply: FastifyReply) => {
    const body = request.body as { params?: Record<string, unknown> };
    const params = body.params ?? {};

    const searchQuery = String(params.query || '').trim();
    const sourceTypes = params.source_types as string[] | undefined;
    const offerId = params.offer_id as string | undefined;
    const categoryId = params.category_id as string | undefined;
    const language = (params.language as string) || 'en';
    const limit = Math.min((params.limit as number) || 10, 50);
    const minScore = (params.min_score as number) || 0.3;

    if (!searchQuery) {
      return reply.send(
        createSuccessResponse({
          chunks: [],
          total_count: 0,
          query: '',
        })
      );
    }

    logger.info({ query: searchQuery, sourceTypes, offerId, limit }, 'Searching knowledge');

    try {
      // Build SQL query
      const conditions: string[] = [];
      const values: unknown[] = [];
      let paramIndex = 1;

      // Text search
      conditions.push(`text ILIKE $${paramIndex}`);
      values.push(`%${searchQuery}%`);
      paramIndex++;

      // Source type filter
      if (sourceTypes && sourceTypes.length > 0) {
        conditions.push(`source_type = ANY($${paramIndex})`);
        values.push(sourceTypes);
        paramIndex++;
      }

      // Offer ID filter
      if (offerId) {
        conditions.push(`offer_id = $${paramIndex}`);
        values.push(offerId);
        paramIndex++;
      }

      // Category ID filter
      if (categoryId) {
        conditions.push(`category_id = $${paramIndex}`);
        values.push(categoryId);
        paramIndex++;
      }

      // Language filter
      conditions.push(`language = $${paramIndex}`);
      values.push(language);
      paramIndex++;

      const whereClause = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : '';

      // Query chunks
      values.push(limit * 3);
      const chunks = await query<EvidenceChunk>(
        `SELECT id, text, source_type, offer_id, sku_id, category_id, language, offsets, created_at
         FROM agent.evidence_chunks
         ${whereClause}
         ORDER BY created_at DESC
         LIMIT $${paramIndex}`,
        values
      );

      // Score and rank results
      const queryTokens = searchQuery.toLowerCase().split(/\s+/);
      const scoredResults = chunks.map(chunk => {
        let score = 0;
        const textLower = chunk.text.toLowerCase();
        
        if (textLower.includes(searchQuery.toLowerCase())) {
          score += 0.5;
        }
        
        for (const token of queryTokens) {
          if (textLower.includes(token)) {
            score += 0.1;
            if (textLower.indexOf(token) < 100) {
              score += 0.05;
            }
          }
        }
        
        score = Math.min(score, 1.0);
        
        return {
          chunk_id: chunk.id,
          text: chunk.text,
          source_type: chunk.source_type,
          offer_id: chunk.offer_id,
          score,
          offsets: chunk.offsets,
          citation: generateCitation(chunk),
        };
      });

      const filteredResults = scoredResults
        .filter(r => r.score >= minScore)
        .sort((a, b) => b.score - a.score)
        .slice(0, limit);

      logger.info({ 
        query: searchQuery, 
        results: filteredResults.length 
      }, 'Knowledge search completed');

      return reply.send(
        createSuccessResponse({
          chunks: filteredResults,
          total_count: filteredResults.length,
          query: searchQuery,
          method: 'keyword',
        }, {
          ttl_seconds: 300,
        })
      );
    } catch (error) {
      logger.error({ error }, 'Knowledge search failed');
      return reply.status(500).send(
        createErrorResponse('INTERNAL_ERROR', 'Failed to search knowledge')
      );
    }
  });

  /**
   * knowledge.get_chunk
   * 
   * Get full chunk content with citation
   */
  app.post('/get_chunk', async (request: FastifyRequest, reply: FastifyReply) => {
    const body = request.body as { params?: { chunk_id?: string } };
    const chunkId = body.params?.chunk_id;

    if (!chunkId) {
      return reply.status(400).send(
        createErrorResponse('INVALID_ARGUMENT', 'chunk_id is required')
      );
    }

    try {
      const chunk = await queryOne<EvidenceChunk>(
        `SELECT * FROM agent.evidence_chunks WHERE id = $1`,
        [chunkId]
      );

      if (!chunk) {
        return reply.status(404).send(
          createErrorResponse('NOT_FOUND', `Chunk ${chunkId} not found`)
        );
      }

      return reply.send(
        createSuccessResponse({
          chunk_id: chunk.id,
          text: chunk.text,
          source_type: chunk.source_type,
          offer_id: chunk.offer_id,
          sku_id: chunk.sku_id,
          category_id: chunk.category_id,
          language: chunk.language,
          offsets: chunk.offsets,
          created_at: chunk.created_at,
          citation: generateCitation(chunk),
        })
      );
    } catch (error) {
      logger.error({ error, chunkId }, 'Failed to get chunk');
      return reply.status(500).send(
        createErrorResponse('INTERNAL_ERROR', 'Failed to get chunk')
      );
    }
  });

  /**
   * knowledge.index_product
   * 
   * Index product content into evidence chunks
   */
  app.post('/index_product', async (request: FastifyRequest, reply: FastifyReply) => {
    const body = request.body as { params?: Record<string, unknown> };
    const params = body.params ?? {};

    const offerId = params.offer_id as string;
    const content = params.content as Record<string, unknown>;
    const sourceType = (params.source_type as string) || 'product_description';
    const language = (params.language as string) || 'en';

    if (!offerId || !content) {
      return reply.status(400).send(
        createErrorResponse('INVALID_ARGUMENT', 'offer_id and content are required')
      );
    }

    logger.info({ offerId }, 'Indexing product');

    try {
      const indexedChunks: string[] = [];

      // Index description
      if (content.description && typeof content.description === 'string') {
        const chunks = chunkText(content.description);
        
        for (const chunk of chunks) {
          const embedding = generateSimpleEmbedding(chunk.text);
          const embeddingStr = `[${embedding.join(',')}]`;
          
          const result = await queryOne<{ id: string }>(
            `INSERT INTO agent.evidence_chunks (text, source_type, offer_id, language, offsets, embedding)
             VALUES ($1, $2, $3, $4, $5, $6::vector)
             RETURNING id`,
            [chunk.text, sourceType, offerId, language, { start: chunk.start, end: chunk.end }, embeddingStr]
          );
          
          if (result) {
            indexedChunks.push(result.id);
          }
        }
      }

      // Index specifications
      if (content.specifications && typeof content.specifications === 'object') {
        const specText = Object.entries(content.specifications as Record<string, unknown>)
          .map(([key, value]) => `${key}: ${value}`)
          .join('\n');
        
        if (specText.length > 0) {
          const embedding = generateSimpleEmbedding(specText);
          const embeddingStr = `[${embedding.join(',')}]`;
          
          const result = await queryOne<{ id: string }>(
            `INSERT INTO agent.evidence_chunks (text, source_type, offer_id, language, embedding)
             VALUES ($1, 'specification', $2, $3, $4::vector)
             RETURNING id`,
            [specText, offerId, language, embeddingStr]
          );
          
          if (result) {
            indexedChunks.push(result.id);
          }
        }
      }

      logger.info({ offerId, chunks: indexedChunks.length }, 'Product indexed');

      return reply.send(
        createSuccessResponse({
          offer_id: offerId,
          indexed_chunks: indexedChunks.length,
          chunk_ids: indexedChunks,
        })
      );
    } catch (error) {
      logger.error({ error, offerId }, 'Failed to index product');
      return reply.status(500).send(
        createErrorResponse('INTERNAL_ERROR', 'Failed to index product')
      );
    }
  });

  /**
   * knowledge.sync_xoobay
   * 
   * Sync XOOBAY products to RAG system
   */
  app.post('/sync_xoobay', async (request: FastifyRequest, reply: FastifyReply) => {
    const body = request.body as { params?: Record<string, unknown> };
    const params = body.params ?? {};

    const pageStart = (params.page_start as number) || 1;
    const pageEnd = (params.page_end as number) || 10;
    const language = (params.language as string) || 'en';

    logger.info({ pageStart, pageEnd }, 'Starting XOOBAY sync');

    const results = {
      pages_processed: 0,
      products_indexed: 0,
      chunks_created: 0,
      errors: [] as string[],
    };

    try {
      const client = getXOOBAYClient();

      for (let page = pageStart; page <= pageEnd; page++) {
        try {
          const productList = await client.getProductList({ pageNo: page, lang: language });
          
          if (!productList.list || productList.list.length === 0) {
            continue;
          }

          for (const product of productList.list) {
            try {
              // Get product details
              const detail = await client.getProductInfo(product.id, language);
              const offerId = `xoobay_${product.id}`;
              
              // Create content
              const description = [
                detail.name,
                detail.short_description,
                detail.description,
              ].filter(Boolean).join('\n\n');

              if (description.length > 0) {
                const chunks = chunkText(description);
                
                for (const chunk of chunks) {
                  const embedding = generateSimpleEmbedding(chunk.text);
                  const embeddingStr = `[${embedding.join(',')}]`;
                  
                  await query(
                    `INSERT INTO agent.evidence_chunks (text, source_type, offer_id, category_id, language, offsets, embedding)
                     VALUES ($1, 'product_description', $2, $3, $4, $5, $6::vector)
                     ON CONFLICT DO NOTHING`,
                    [chunk.text, offerId, detail.category || 'other', language, { start: chunk.start, end: chunk.end }, embeddingStr]
                  );
                  
                  results.chunks_created++;
                }
              }

              results.products_indexed++;
            } catch (productError) {
              results.errors.push(`Product ${product.id}: ${productError instanceof Error ? productError.message : String(productError)}`);
            }
          }

          results.pages_processed++;
          
          // Check circuit breaker status
          const status = client.getCircuitStatus();
          if (status.state === 'OPEN') {
            logger.warn('Circuit breaker open, stopping sync');
            break;
          }
        } catch (pageError) {
          results.errors.push(`Page ${page}: ${pageError instanceof Error ? pageError.message : String(pageError)}`);
        }
      }

      logger.info(results, 'XOOBAY sync completed');

      return reply.send(
        createSuccessResponse(results)
      );
    } catch (error) {
      logger.error({ error }, 'XOOBAY sync failed');
      return reply.status(500).send(
        createErrorResponse('INTERNAL_ERROR', 'Failed to sync XOOBAY products')
      );
    }
  });

  /**
   * knowledge.stats
   * 
   * Get knowledge base statistics
   */
  app.post('/stats', async (_request: FastifyRequest, reply: FastifyReply) => {
    try {
      const stats = await queryOne<{
        total_chunks: string;
        source_types: Record<string, number>;
        languages: Record<string, number>;
      }>(
        `SELECT 
          COUNT(*) as total_chunks,
          jsonb_object_agg(source_type, count) as source_types,
          jsonb_object_agg(language, lang_count) as languages
         FROM (
           SELECT source_type, COUNT(*) as count
           FROM agent.evidence_chunks
           GROUP BY source_type
         ) st,
         (
           SELECT language, COUNT(*) as lang_count
           FROM agent.evidence_chunks
           GROUP BY language
         ) lg`
      );

      // Get XOOBAY client status
      const xoobayClient = getXOOBAYClient();
      const circuitStatus = xoobayClient.getCircuitStatus();

      return reply.send(
        createSuccessResponse({
          total_chunks: parseInt(stats?.total_chunks || '0'),
          source_types: stats?.source_types || {},
          languages: stats?.languages || {},
          xoobay_circuit: circuitStatus,
        })
      );
    } catch (error) {
      logger.error({ error }, 'Failed to get stats');
      return reply.status(500).send(
        createErrorResponse('INTERNAL_ERROR', 'Failed to get knowledge stats')
      );
    }
  });
}

File: agents/src/tools/knowledge.py
"""
Knowledge tools - RAG 检索
"""

from typing import Any

from .base import MOCK_MODE, call_tool, mock_response


async def search_knowledge(
    query: str,
    source_types: list[str] | None = None,
    offer_id: str | None = None,
    category_id: str | None = None,
    language: str = "en",
    limit: int = 10,
    min_score: float = 0.3,
    user_id: str | None = None,
) -> dict[str, Any]:
    """
    knowledge.search - 搜索知识库

    Args:
        query: 搜索查询
        source_types: 来源类型过滤
        offer_id: 按商品 ID 过滤
        category_id: 按类目过滤
        language: 语言
        limit: 最大结果数
        min_score: 最低相关性分数

    Returns:
        标准响应 Envelope，data 包含检索结果
    """
    if MOCK_MODE:
        return mock_response({
            "chunks": [
                {
                    "chunk_id": "chunk_mock_001",
                    "text": f"This is mock evidence about {query}.",
                    "source_type": "product_description",
                    "offer_id": offer_id,
                    "score": 0.85,
                    "citation": "[chunk:chunk_mock_001]",
                },
            ],
            "total_count": 1,
            "query": query,
            "method": "mock",
        })

    return await call_tool(
        mcp_server="core",
        tool_name="knowledge.search",
        params={
            "query": query,
            "source_types": source_types,
            "offer_id": offer_id,
            "category_id": category_id,
            "language": language,
            "limit": limit,
            "min_score": min_score,
        },
        user_id=user_id,
    )


async def get_chunk(
    chunk_id: str,
    user_id: str | None = None,
) -> dict[str, Any]:
    """
    knowledge.get_chunk - 获取证据块

    Args:
        chunk_id: 证据块 ID

    Returns:
        标准响应 Envelope，data 包含完整证据块
    """
    if MOCK_MODE:
        return mock_response({
            "chunk_id": chunk_id,
            "text": "This is the full content of the evidence chunk.",
            "source_type": "product_description",
            "offer_id": None,
            "offsets": {"start": 0, "end": 100},
            "citation": f"[chunk:{chunk_id}]",
        })

    return await call_tool(
        mcp_server="core",
        tool_name="knowledge.get_chunk",
        params={"chunk_id": chunk_id},
        user_id=user_id,
    )


async def index_product(
    offer_id: str,
    content: dict,
    source_type: str = "product_description",
    language: str = "en",
    user_id: str | None = None,
) -> dict[str, Any]:
    """
    knowledge.index_product - 索引产品内容

    Args:
        offer_id: 商品 ID
        content: 要索引的内容
        source_type: 来源类型
        language: 语言

    Returns:
        标准响应 Envelope，data 包含索引结果
    """
    if MOCK_MODE:
        return mock_response({
            "offer_id": offer_id,
            "indexed_chunks": 1,
            "chunk_ids": ["chunk_mock_new"],
        })

    return await call_tool(
        mcp_server="core",
        tool_name="knowledge.index_product",
        params={
            "offer_id": offer_id,
            "content": content,
            "source_type": source_type,
            "language": language,
        },
        user_id=user_id,
    )


async def search_with_context(
    query: str,
    offer_ids: list[str] | None = None,
    include_compliance: bool = True,
    include_shipping: bool = True,
    language: str = "en",
    user_id: str | None = None,
) -> dict[str, Any]:
    """
    综合知识搜索 - 同时搜索产品、合规和物流信息

    Args:
        query: 搜索查询
        offer_ids: 限定的商品 ID 列表
        include_compliance: 是否包含合规信息
        include_shipping: 是否包含物流信息
        language: 语言

    Returns:
        组合的搜索结果
    """
    results = {
        "product_chunks": [],
        "compliance_chunks": [],
        "shipping_chunks": [],
        "total_chunks": 0,
    }

    # 搜索产品信息
    product_result = await search_knowledge(
        query=query,
        source_types=["product_description", "specification", "qa"],
        language=language,
        limit=5,
        user_id=user_id,
    )
    if product_result.get("ok"):
        results["product_chunks"] = product_result.get("data", {}).get("chunks", [])

    # 搜索合规信息
    if include_compliance:
        compliance_result = await search_knowledge(
            query=f"{query} compliance certification",
            source_types=["compliance", "policy"],
            language=language,
            limit=3,
            user_id=user_id,
        )
        if compliance_result.get("ok"):
            results["compliance_chunks"] = compliance_result.get("data", {}).get("chunks", [])

    # 搜索物流信息
    if include_shipping:
        shipping_result = await search_knowledge(
            query=f"{query} shipping delivery",
            source_types=["shipping", "policy"],
            language=language,
            limit=3,
            user_id=user_id,
        )
        if shipping_result.get("ok"):
            results["shipping_chunks"] = shipping_result.get("data", {}).get("chunks", [])

    # 计算总数
    results["total_chunks"] = (
        len(results["product_chunks"]) +
        len(results["compliance_chunks"]) +
        len(results["shipping_chunks"])
    )

    return {"ok": True, "data": results}


